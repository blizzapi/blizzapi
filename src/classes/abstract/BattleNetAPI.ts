import { OAuth2API } from './OAuth2API';
import * as OAuthHelpers from '../../helpers/oauth';
import {
  RegionIdOrName,
  AccessToken,
  BattleNetOptions,
} from '../../types';

export abstract class BattleNetAPI extends OAuth2API {
  readonly region: RegionIdOrName;

  accessToken: AccessToken;

  constructor(options: BattleNetOptions) {
    super(options.clientId, options.clientSecret);
    this.region = options.region;
    this.accessToken = options.accessToken || undefined;
  }

  getAccessToken = () => (this.accessToken || this.setAccessToken());

  setAccessToken = async () => {
    this.accessToken = await OAuthHelpers.getAccessToken({
      region: this.region,
      clientId: this.clientId,
      clientSecret: this.clientSecret,
    });
    return this.accessToken;
  };

  static validateAccessToken = (region: RegionIdOrName, accessToken: AccessToken) =>
    OAuthHelpers.validateAccessToken(region, accessToken);
}
