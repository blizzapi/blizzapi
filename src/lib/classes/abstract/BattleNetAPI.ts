import { ClientId, ClientSecret, RegionIdOrName, AccessToken } from '../../../../@types';
import OAuth2API from './OAuth2API';
import * as OAuthHelpers from '../../helpers/oauth';

/* tslint:disable:no-this no-expression-statement no-object-mutation readonly-keyword */

export = class BattleNetAPI extends OAuth2API {
  readonly region: RegionIdOrName;
  accessToken: AccessToken;

  constructor(
    region: RegionIdOrName,
    clientId: ClientId,
    clientSecret: ClientSecret,
    accessToken?: AccessToken,
  ) {
    super(clientId, clientSecret);
    this.region = region;
    this.accessToken = accessToken || undefined;
  }

  readonly getAccessToken = () => (this.accessToken ? this.accessToken : this.setAccessToken());

  readonly setAccessToken = async () =>
    (this.accessToken = await OAuthHelpers.getAccessToken(
      this.region,
      this.clientId,
      this.clientSecret,
    ));

  static validateAccessToken = async (region: RegionIdOrName, accessToken: AccessToken) =>
    OAuthHelpers.validateAccessToken(region, accessToken);
};

/* tslint:enable:no-this no-expression-statement no-object-mutation readonly-keyword */
