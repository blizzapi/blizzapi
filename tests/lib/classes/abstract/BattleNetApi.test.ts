import BattleNetAPI from '../../../../src/lib/classes/abstract/BattleNetAPI';
// import * as oauth from '../../../../src/lib/helpers';

jest.genMockFromModule('../../../../src/lib/classes/abstract/BattleNetAPI');

// tslint:disable-next-line: no-object-mutation
BattleNetAPI.prototype.getAccessToken = jest.fn().mockImplementation(() =>
  Promise.resolve('valid_access_token_from_BattleNetAPIMock'));

// tslint:disable
class BattleNetAPIMock extends BattleNetAPI {
  constructor(accessToken?: string) {
    super({
      accessToken,
      region: 'us',
      clientId: 'valid_client_id_from_BattleNetAPIMock',
      clientSecret: 'valid_client_secret_from_BattleNetAPIMock',
    });
  }

  query(uri: string) {
    return {
      status: 'ok',
      uri,
    };
  }
}
// tslint:enable

describe('BattleNetAPI class', () => {
  test('should be defined', () => {
    expect(BattleNetAPI).toBeDefined();
  });

  test('should define region property', () => {
    expect(new BattleNetAPIMock().region).toMatchSnapshot();
  });

  test('should define clientId property', () => {
    expect(new BattleNetAPIMock().clientId).toMatchSnapshot();
  });

  test('should define clientSecret property', () => {
    expect(new BattleNetAPIMock().clientSecret).toMatchSnapshot();
  });

  test('should save accessToken as undefined', () => {
    expect(new BattleNetAPIMock().accessToken).toMatchSnapshot();
  });

  test('should set access token', async () => {
    const bnetApi = new BattleNetAPIMock();
    const response = await bnetApi.setAccessToken();
    expect(response).toMatchSnapshot();
  });

  test('should get access token', async () => {
    const bnetApi = new BattleNetAPIMock();
    const response = await bnetApi.getAccessToken();
    expect(response).toMatchSnapshot();
  });


  test('should get access token when provided by user', async () => {
    const bnetApi = new BattleNetAPIMock('valid_access_token');
    const response = await bnetApi.getAccessToken();
    expect(response).toMatchSnapshot();
  });

  test('should validate access token and return true', async () => {
    const response = await BattleNetAPIMock.validateAccessToken('us', 'valid_access_token');
    expect(response).toMatchSnapshot();
  });
});