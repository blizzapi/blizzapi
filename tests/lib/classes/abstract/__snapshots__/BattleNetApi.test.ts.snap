// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BattleNetAPI class should define clientId property 1`] = `"valid_client_id_from_BattleNetAPIMock"`;

exports[`BattleNetAPI class should define clientSecret property 1`] = `"valid_client_secret_from_BattleNetAPIMock"`;

exports[`BattleNetAPI class should define region property 1`] = `"us"`;

exports[`BattleNetAPI class should get access token 1`] = `"dmFsaWRfY2xpZW50X2lkX2Zyb21fQmF0dGxlTmV0QVBJTW9jazp2YWxpZF9jbGllbnRfc2VjcmV0X2Zyb21fQmF0dGxlTmV0QVBJTW9jaw=="`;

exports[`BattleNetAPI class should get access token when provided by user 1`] = `"valid_access_token"`;

exports[`BattleNetAPI class should save accessToken as undefined 1`] = `undefined`;

exports[`BattleNetAPI class should set access token 1`] = `"dmFsaWRfY2xpZW50X2lkX2Zyb21fQmF0dGxlTmV0QVBJTW9jazp2YWxpZF9jbGllbnRfc2VjcmV0X2Zyb21fQmF0dGxlTmV0QVBJTW9jaw=="`;

exports[`BattleNetAPI class should validate access token and return true 1`] = `true`;
